lib/request/RequestQueueV2.js

```diff
@@ -63,36 +66,39 @@ RequestQueueV2.prototype = {
             // The request has been sent, attempt to jump on the request
             // if possible.
             if (request.sent) {
-                var results = request.add(rRemainingPaths, oRemainingPaths, refCountCallback);
-
-                // Checks to see if the results were successfully inserted
-                // into the outgoing results.  Then our paths will be reduced
-                // to the complement.
-                if (results[0]) {
-                    rRemainingPaths = results[1];
-                    oRemainingPaths = results[2];
-                    disposables[disposables.length] = results[3];
-                    ++count;
+                if (this.model._enableRequestDeduplication) {
+                    var results = request.add(rRemainingPaths, oRemainingPaths, refCountCallback);
+
+                    // Checks to see if the results were successfully inserted
+                    // into the outgoing results.  Then our paths will be reduced
+                    // to the complement.
+                    if (results[0]) {
+                        rRemainingPaths = results[1];
+                        oRemainingPaths = results[2];
+                        disposables[disposables.length] = results[3];
+                        ++count;
+
+                        // If there are no more remaining paths then exit the loop.
+                        if (!oRemainingPaths.length) {
+                            break;
+                        }
+                    }
                 }
             }
 
-            // If there is a non sent request, then we can batch and leave.
+            // If there is an unsent request, then we can batch and leave.
             else {
                 request.batch(rRemainingPaths, oRemainingPaths, refCountCallback);
-                oRemainingPaths = [];
-                rRemainingPaths = [];
+                oRemainingPaths = null;
```

Comment:
i understand that the reasoning behind doing this is to reduce garbage creation?

Reply:
Correct. The arrays are not used afterwards except to check they're empty, so these allocations were unnecessary.